### ISSUE TITLE
model.save() fails when model has been populated

### ISSUE THREAD

**Issue by lwansbrough** [balderdashy/sails#2073]
I'm getting a Waterline error when I try to re-save a populated model. The association I'm populating is one-to-many on the model being selected. Here's the WL error:

    {"error":"E_VALIDATION","status":400,"summary":"1 attribute is invalid","model":"MatchReport","invalidAttributes":{"match":[{"rule":"required","message":"\"required\" validation rule failed for input: null"},{"rule":"notNull","message":"\"notNull\" validation rule failed for input: null"}]}}

So the problem appears to be that `match` on `MatchReport` is undefined/null. The model being populated is a `Match` with a collection of `MatchReport` - seems like Waterline is tripping up on trying to prevent a circular reference or something?

To clarify further, here's the code:

    Match.findOne(id)
       .populate('reports')
       .exec(function(err, match) {
         if(err) return done(err);
         match.status = 'DISPUTE';
         match.save(function(err) {
           if(err) return done(err); // Error happens here.
           ...
         });
       });

When I inspect `match` the `reports` is `undefined` and when I inspect `match.reports` the `reports` are shown correctly, and `match.reports[0].match` is a UUID as expected - so it's not as though they're failing to populate.

**Comment by UzEE** [comment]
The workaround wouldn't work in case you want to update both the main model and the associated model in one go.

Example:

```js
Product
	.findOne({id: spotId})
	.populate('video')
	.exec(function spotFound(err, spot) {

		/* ... */

		spot.status = 'accepted';
		spot.video.status = 'ready';

		spot.save(function spotSaved(err, updatedSpot) {
			/* ... */
		});
	});
```

Which throws:

```
Error (E_VALIDATION) :: 1 attribute is invalid
    at WLValidationError.WLError (app\node_modules\sails\node_modules\waterline\lib\waterline\error\WLError.js:33:18)
    at new WLValidationError (app\node_modules\sails\node_modules\waterline\lib\waterline\error\WLValidationError.js:20:28)
    at app\node_modules\sails\node_modules\waterline\lib\waterline\query\validate.js:45:43
    at allValidationsChecked (app\node_modules\sails\node_modules\waterline\lib\waterline\core\validations.js:203:5)
    at done (app\node_modules\sails\node_modules\async\lib\async.js:135:19)
    at app\node_modules\sails\node_modules\async\lib\async.js:32:16
    at app\node_modules\sails\node_modules\waterline\lib\waterline\core\validations.js:194:14
    at done (app\node_modules\sails\node_modules\async\lib\async.js:135:19)
    at app\node_modules\sails\node_modules\async\lib\async.js:32:16
    at app\node_modules\sails\node_modules\waterline\lib\waterline\core\validations.js:157:64
    at app\node_modules\sails\node_modules\async\lib\async.js:125:13
    at Array.forEach (native)
    at _each (app\node_modules\sails\node_modules\async\lib\async.js:46:24)
    at Object.async.each (app\node_modules\sails\node_modules\async\lib\async.js:124:9)
    at validate (app\node_modules\sails\node_modules\waterline\lib\waterline\core\validations.js:156:11)
    at app\node_modules\sails\node_modules\async\lib\async.js:125:13

Invalid attributes sent to Product:
 • video
   • "required" validation rule failed for input: null
```

The weird problem is that it only happens one time when you try to save the object through this logic. Immediately calling the route endpoint again works correctly without throwing any errors.

**Comment by sgress454** [comment]
@UzEE this should be fixed...if you're still seeing it please make an issue in the [Waterline repo](https://github.com/balderdashy/waterline)

**Comment by drbolsen** [comment]
the same issue as @UzEE: updating both a main model and an association in one go will return an error in callback and the result value is empty however both models are actually updated. Weird.

**Comment by drbolsen** [comment]
I have experimented a bit and found that by adding the following line to my code solves the problem:
```
passport.password = req.body.password;
passport.user = passport.user.id;  <---- This fixes the issue in my case
passport.save(function (err, saved) {
```
I thought it wouldn't be necessary as this was already part of the populated collection
```json
{"user":{
"username":"Tester",
"email":"Tester@test.com",
"resetPasswordToken":"76bb160e-cb57-40ac-829c-9d42584ea3f8",
"resetPasswordExpires":"2015-04-25T12:28:28.154Z",
"title":"Mr",
"id":1,
"createdAt":"2015-04-25T10:15:09.795Z",
"updatedAt":"2015-04-25T11:28:28.218Z"
},
"protocol":"local",
"password":"$2a$10$8AdCKFwAUKR0xGElMUFsLefNHn4BtvSu1Ch/LVFLxIx.j7iQqE9O6",
"accessToken":"LRNfpDFTzwTT3WEEpx2K2dPxbo2ePgbmtIkn8OcKgi/R9jrqWqA2I0Gf4ywvSAV8",
"id":1,
"createdAt":"2015-04-25T10:15:10.077Z",
"updatedAt":"2015-04-25T11:24:48.398Z"}
```
but it appears that an association attribute that links two models must be specified.

**Comment by drbolsen** [comment]
Following up on this one, further digging into the issue I found that my fix is not actually a fix - it just mitigates an apparent exception error message. In reality assigning "id", which is in essence a foreign key that links two models, will just re-link the populated association to the updated record. No changes are made. So in my particular scenario - resetPasswordToken will never be updated.  But it seems like an expected behaviour though. And I have no issues with this.

My original issue - an exception error  __"required" validation rule failed for input: null"__ is a bit different, I looked into sources and found that it had been caused by this piece of code:

sails/node_modules/waterline/lib/waterline/utils/nestedOperations/update.js

```javascript

  438      // Run an update operation to set the foreign key to null on all the
  439      // associated child records.
  440:     values[operation.keyName] = null; <---- This will set an attribute to null
  441      
  442      self.waterline.collections[operation.model].update(operation.criteria, values).exec(next);

```

Basically if a key set to __required__ like in my model:

User.js

```javascript
var User = {
  // Enforce model schema in the case of schemaless databases
  schema: true,

  attributes: {
    username  : { type: 'string', unique: true },
    email     : { type: 'email',  unique: true },
    passports : { collection: 'Passport', via: 'user' }
  }
};
```

Passport.js

```javascript
var Passport = {
  attributes: {

    protocol: { type: 'alphanumeric', required: true },

    password    : { type: 'string', minLength: 8 },
    accessToken : { type: 'string' },


    provider   : { type: 'alphanumericdashed' },
    identifier : { type: 'string' },
    tokens     : { type: 'json' },

    user: { model: 'User', required: true },  <-------- Here 

    validatePassword: function (password, next) {
      bcrypt.compare(password, this.password, next);
    }
  },
```

It will trigger an exception because the update happens in few rounds - first round updates all associated child records, the next unlink them by setting foreign keys to __null__ and then reconnect them again:

```javascript
function sync(parents, operations, cb) {
  var self = this;

  async.auto({

    // Update any nested associations
    update: function(next) {
      updateRunner.call(self, parents, operations, next);
    },

    // For each parent, unlink all the associations currently set
    unlink: ['update', function(next) {
      unlinkRunner.call(self, parents, operations, next);
    }],

    // For each parent found, link any associations passed in by either creating
    // the new record or linking an existing record
    link: ['unlink', function(next) {
      linkRunner.call(self, parents, operations, next);
    }]

  }, cb);
}

```

That is why I had my models updated regardless of exception.

**Comment by mikermcneil** [comment]
Thanks for posting, @lwansbrough.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
